File: source/characters.c
#include <stddef.h>
#include "characters.h"
#include "assets.h"   /* bin2s-generated symbols for every embedded asset */

/* GENERATED -- do not hand-edit. Mirrors the source game:
 *   hitbox_*        <- the `hitbox` block in js/characters/<id>.js
 *   thrust/gravity/ <- js/character-balance.js
 *   maxVelocity
 *   sprite_origin_* <- measured by re-rendering each draw() and locating
 *                      the art inside its 128x128 export; differs per
 *                      character, so it cannot be a shared constant.
 * Sprite bytes are linked into the binary by the Makefile's bin2s rule; the
 * symbols come from assets.h. Roster order matches the web game's load order. */
static const CharacterData all_characters[] = {
    {
        "cat-synth", "Synth Cat",
        cat_synth_idle_png,
        cat_synth_thrust_png,
        42, 30, -1, 18,
        44, 44,
        -0.6f, 0.36f, 10.5f
    },
    {
        "beava", "Beava",
        beava_idle_png,
        beava_thrust_png,
        40, 35, 0, 0,
        37, 48,
        -0.6f, 0.4f, 10.0f
    },
    {
        "vinny-bobarino", "Vinny Bobarino",
        vinny_bobarino_idle_png,
        vinny_bobarino_thrust_png,
        35, 35, 15, 0,
        45, 44,
        -0.6f, 0.4f, 10.0f
    },
    {
        "darius-hodgekins", "Darius Hodgekins",
        darius_hodgekins_idle_png,
        darius_hodgekins_thrust_png,
        40, 35, 0, 0,
        54, 42,
        -0.6f, 0.4f, 10.0f
    },
    {
        "roderick-tron", "Roderick Tron",
        roderick_tron_idle_png,
        roderick_tron_thrust_png,
        40, 35, 0, 0,
        55, 39,
        -0.6f, 0.4f, 10.0f
    },
    {
        "juanito-thompson", "Juanito Thompson",
        juanito_thompson_idle_png,
        juanito_thompson_thrust_png,
        40, 35, 0, 0,
        46, 54,
        -0.6f, 0.4f, 10.0f
    },
    {
        "carl-spatski", "Carl Spatski",
        carl_spatski_idle_png,
        carl_spatski_thrust_png,
        42, 42, -1, -2,
        58, 44,
        -0.65f, 0.4f, 10.5f
    },
    {
        "dspum-balloon", "dspum balloon",
        dspum_balloon_idle_png,
        dspum_balloon_thrust_png,
        30, 38, 5, -3,
        44, 44,
        -0.5f, 0.3f, 9.0f
    },
    {
        "foresters-soul", "Forester's Soul",
        foresters_soul_idle_png,
        foresters_soul_thrust_png,
        40, 40, 0, -2,
        59, 44,
        -0.62f, 0.38f, 11.0f
    },
    {
        "grocery-harrison", "Grocery Harrison",
        grocery_harrison_idle_png,
        grocery_harrison_thrust_png,
        35, 30, 8, 5,
        54, 46,
        -0.6f, 0.4f, 10.0f
    },
    {
        "didgeridoo-man", "Didgeridoo Man",
        didgeridoo_man_idle_png,
        didgeridoo_man_thrust_png,
        45, 35, 0, 0,
        23, 38,
        -0.7f, 0.35f, 9.5f
    },
    {
        "mountain-gnome", "Mountain Gnome",
        mountain_gnome_idle_png,
        mountain_gnome_thrust_png,
        32, 32, 4, 0,
        44, 44,
        -0.55f, 0.32f, 9.5f
    },
    {
        "prince-vince", "Prince Vince",
        prince_vince_idle_png,
        prince_vince_thrust_png,
        27, 38, 7, -3,
        60, 43,
        -0.65f, 0.38f, 10.5f
    },
    {
        "dag-henderson", "Dag Henderson",
        dag_henderson_idle_png,
        dag_henderson_thrust_png,
        40, 35, 0, 0,
        48, 44,
        -0.6f, 0.4f, 10.0f
    },
    {
        "ban-daniel", "BANDANIEL",
        ban_daniel_idle_png,
        ban_daniel_thrust_png,
        38, 38, 1, -2,
        40, 44,
        -0.6f, 0.4f, 10.0f
    },
    {
        "plantain-jane", "PLANTAIN JANE",
        plantain_jane_idle_png,
        plantain_jane_thrust_png,
        38, 38, 1, -2,
        42, 44,
        -0.6f, 0.4f, 10.0f
    },
    {
        "fire-toad", "Fire Toad",
        fire_toad_idle_png,
        fire_toad_thrust_png,
        32, 32, 4, 0,
        35, 44,
        -0.65f, 0.38f, 10.5f
    },
    {
        "backpack-man", "Backpack Man",
        backpack_man_idle_png,
        backpack_man_thrust_png,
        34, 36, 6, -2,
        44, 44,
        -0.6f, 0.4f, 10.0f
    },
    {
        "tollbooth-lady", "Tollbooth Lady",
        tollbooth_lady_idle_png,
        tollbooth_lady_thrust_png,
        40, 35, 0, 0,
        56, 44,
        -0.6f, 0.4f, 10.0f
    },
    {
        "svfp-van", "SVFP Van",
        svfp_van_idle_png,
        svfp_van_thrust_png,
        36, 18, 4, 12,
        52, 42,
        -0.6f, 0.4f, 10.0f
    },
    {
        "tardigrade", "Tardigrade",
        tardigrade_idle_png,
        tardigrade_thrust_png,
        26, 25, 7, 8,
        51, 45,
        -0.7f, 0.35f, 11.0f
    },
    {
        "elektra", "Elektra",
        elektra_idle_png,
        elektra_thrust_png,
        38, 38, 1, 0,
        44, 44,
        -0.58f, 0.38f, 10.5f
    },
    {
        "strawberto", "Strawberto",
        strawberto_idle_png,
        strawberto_thrust_png,
        36, 38, 2, 0,
        44, 37,
        -0.6f, 0.39f, 10.0f
    },
    {
        "carl", "Carl",
        carl_idle_png,
        carl_thrust_png,
        38, 38, 1, 0,
        49, 39,
        -0.62f, 0.37f, 11.0f
    }
};

#define CHARACTER_COUNT (sizeof(all_characters) / sizeof(all_characters[0]))

static int current_index = 0;

void characters_init(void) {
    current_index = 0;
}

const CharacterData *characters_get_current(void) {
    return &all_characters[current_index];
}

void characters_set_current(int index) {
    if (index >= 0 && index < (int)CHARACTER_COUNT) {
        current_index = index;
    }
}

int characters_get_current_index(void) {
    return current_index;
}

int characters_get_count(void) {
    return (int)CHARACTER_COUNT;
}

const CharacterData *characters_get_by_index(int index) {
    if (index >= 0 && index < (int)CHARACTER_COUNT) {
        return &all_characters[index];
    }
    return NULL;
}

File: tui/drift/characters.py
"""The roster: who you fly as, and how differently they fly.

The physics — ``thrust``, ``gravity`` and ``max_velocity`` — are lifted
verbatim from ``wii/source/characters.c``, which took them from the
``js/characters/`` modules. They are the part of a character that a terminal
can carry perfectly, and the part that actually matters: a heavier character
falls faster and climbs slower, and you feel it in the first second.

**The art cannot come across, and pretending otherwise would be worse.** On the
web each of these is drawn with canvas primitives — gradients, hue cycling, a
bobbing offset — and on the Wii they are PNGs baked into the binary. At the
scale a terminal gives, the player is about two cells wide and one tall (see
:mod:`drift.projection`), and ``draw_image`` is a documented no-op in the
engine's terminal backend anyway. So a character here is three things: a glyph
small enough to fit, an accent colour taken from its web palette, and its
physics.

The hitboxes come across exactly, because collision is in world space and has
nothing to do with what is drawn.
"""

from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class Character:
    """One pilot. Frozen — a run mutates a :class:`~drift.player.Player`."""

    key: str
    name: str
    #: What the player is drawn as. Three cells at most; a wider one would be
    #: bigger than the hitbox it stands for, which is a lie the player pays for.
    glyph: str
    accent: str
    #: ``(w, h, offset_x, offset_y)`` in world units, from the browser build.
    hitbox: tuple[int, int, int, int]
    thrust: float
    gravity: float
    max_velocity: float

    def apply(self, player) -> None:
        """Put this character's physics and hitbox on a player."""
        player.set_physics(self.thrust, self.gravity, self.max_velocity)
        player.set_hitbox(*self.hitbox)


#: Roster order matches the browser build's load order, which is the order the
#: Wii port carries too. Three versions, one order — a character's number is
#: the same wherever you play.
ROSTER: tuple[Character, ...] = (
    Character("cat-synth", "Synth Cat", '=^=', "#ff4fd8",
              hitbox=(42, 30, -1, 18),
              thrust=-0.6, gravity=0.36, max_velocity=10.5),
    Character("beava", "Beava", '<@>', "#c98a3a",
              hitbox=(40, 35, 0, 0),
              thrust=-0.6, gravity=0.4, max_velocity=10.0),
    Character("vinny-bobarino", "Vinny Bobarino", '<D>', "#ffb347",
              hitbox=(35, 35, 15, 0),
              thrust=-0.6, gravity=0.4, max_velocity=10.0),
    Character("darius-hodgekins", "Darius Hodgekins", '<H>', "#8ad3ff",
              hitbox=(40, 35, 0, 0),
              thrust=-0.6, gravity=0.4, max_velocity=10.0),
    Character("roderick-tron", "Roderick Tron", '[T]', "#39ff6e",
              hitbox=(40, 35, 0, 0),
              thrust=-0.6, gravity=0.4, max_velocity=10.0),
    Character("juanito-thompson", "Juanito Thompson", '<J>', "#ff7043",
              hitbox=(40, 35, 0, 0),
              thrust=-0.6, gravity=0.4, max_velocity=10.0),
    Character("carl-spatski", "Carl Spatski", '<%>', "#b388ff",
              hitbox=(42, 42, -1, -2),
              thrust=-0.65, gravity=0.4, max_velocity=10.5),
    Character("dspum-balloon", "dspum balloon", '(o)', "#ff5fa2",
              hitbox=(30, 38, 5, -3),
              thrust=-0.5, gravity=0.3, max_velocity=9.0),
    Character("foresters-soul", "Forester's Soul", '}|{', "#5fe08a",
              hitbox=(40, 40, 0, -2),
              thrust=-0.62, gravity=0.38, max_velocity=11.0),
    Character("grocery-harrison", "Grocery Harrison", '[G]', "#ffd54f",
              hitbox=(35, 30, 8, 5),
              thrust=-0.6, gravity=0.4, max_velocity=10.0),
    Character("didgeridoo-man", "Didgeridoo Man", '==D', "#d2691e",
              hitbox=(45, 35, 0, 0),
              thrust=-0.7, gravity=0.35, max_velocity=9.5),
    Character("mountain-gnome", "Mountain Gnome", '/A\\', "#9ccc65",
              hitbox=(32, 32, 4, 0),
              thrust=-0.55, gravity=0.32, max_velocity=9.5),
    Character("prince-vince", "Prince Vince", '<P>', "#7e57c2",
              hitbox=(27, 38, 7, -3),
              thrust=-0.65, gravity=0.38, max_velocity=10.5),
    Character("dag-henderson", "Dag Henderson", '<b>', "#4fc3f7",
              hitbox=(40, 35, 0, 0),
              thrust=-0.6, gravity=0.4, max_velocity=10.0),
    Character("ban-daniel", "BANDANIEL", '<B>', "#ff1744",
              hitbox=(38, 38, 1, -2),
              thrust=-0.6, gravity=0.4, max_velocity=10.0),
    Character("plantain-jane", "PLANTAIN JANE", '<J>', "#cddc39",
              hitbox=(38, 38, 1, -2),
              thrust=-0.6, gravity=0.4, max_velocity=10.0),
    Character("fire-toad", "Fire Toad", '<*>', "#ff6d00",
              hitbox=(32, 32, 4, 0),
              thrust=-0.65, gravity=0.38, max_velocity=10.5),
    Character("backpack-man", "Backpack Man", '[M]', "#8d6e63",
              hitbox=(34, 36, 6, -2),
              thrust=-0.6, gravity=0.4, max_velocity=10.0),
    Character("tollbooth-lady", "Tollbooth Lady", '<T>', "#f06292",
              hitbox=(40, 35, 0, 0),
              thrust=-0.6, gravity=0.4, max_velocity=10.0),
    Character("svfp-van", "SVFP Van", '[V]', "#26c6da",
              hitbox=(36, 18, 4, 12),
              thrust=-0.6, gravity=0.4, max_velocity=10.0),
    Character("tardigrade", "Tardigrade", '(~)', "#a1887f",
              hitbox=(26, 25, 7, 8),
              thrust=-0.7, gravity=0.35, max_velocity=11.0),
    Character("elektra", "Elektra", '<E>', "#00e5ff",
              hitbox=(38, 38, 1, 0),
              thrust=-0.58, gravity=0.38, max_velocity=10.5),
    Character("strawberto", "Strawberto", '<o>', "#ff3d71",
              hitbox=(36, 38, 2, 0),
              thrust=-0.6, gravity=0.39, max_velocity=10.0),
    Character("carl", "Carl", '<@>', "#ffee58",
              hitbox=(38, 38, 1, 0),
              thrust=-0.62, gravity=0.37, max_velocity=11.0),
)

BY_KEY = {c.key: c for c in ROSTER}

#: Who you fly as before choosing. The browser opens on the same one.
DEFAULT = ROSTER[0]


def get(key: str) -> Character:
    """The named character, or :data:`DEFAULT` if there is no such name."""
    return BY_KEY.get(key, DEFAULT)


__all__ = ["BY_KEY", "DEFAULT", "ROSTER", "Character", "get"]
